### PR TITLE
R&D Techwebs

### PR DESCRIPTION
**Pull request by AffectedArc07**
## What Does This PR Do
**THIS PR DOES NOT CONTAIN DEPARTMENT PROTOLATHES**

This PR adds R&D techwebs as a new research system. You gain points over time (this can be accelerated, see below) using R&D servers, and research nodes as time progresses. This requires science to actually think about what they want to research and stops R&D being gamed within 20 minutes. It turns science into something which actually requires thought, as opposed to "grind through a shopping list for 20 minutes then slack about"

As well as waiting for point income, you can also generate point boosts using anomaly cores, toxins bombs, or chemistry. A new machine (chem consumer) will automatically pick a complex chemical each round, and units of that reagent will get you points. 

Toxins bombs will grant more research points the closer you are to the bomb target.

The R&D server controller now controls console Read/Write access, as well as server status

I actually started this PR over 4 months ago but sidelined it since it required map edits, but now boxstation is moving its coming up. 

## Why It's Good For The Game
R&D has been in an awful state for a while, we discussed this internally and feel this is the best option.

## Images of changes
R&D main menu
image

Techwebs main menu (Shows node status)
image
You can hover over the design images to show what the design gives you

Node view (Shows requirements, the node itself, and what it is a component of)
image

R&D server console main menu
image

R&D server console control menu
image

R&D server console server status
image

Chem consumer
image

Chem consumer UI
image

## Changelog
:cl: AffectedArc07
add: Added R&D techwebs
del: Removed old R&D
add: Research chemical consumer. Turn high-level chems into research points
tweak: Toxins doppler array now generates research points based on explosion target
/:cl:

### CODE REVIEW

**Comment by FlattestGuitar** [comment]
>  It turns science into something which actually requires thought

image

**Comment by shmabuu** [comment]
I’ve personally played around with tech webs a lot on other servers and found it to be a lot more interactive than what we currently have. I did rnd on para for quite a while and sure, this won’t change the fact that there will be an ‘optimal path’, but it will certainly help the fact that it’s always the same path with not much space for lenience. This allows and encourages rnd to go for different things based on current situations and requests.

**Comment by S34NW** [comment]
Does hovering over the circuit boards show what they are? Or will it just be a load of identical icons?

File: code/game/gamemodes/objective.dm [hunk 1]
```diff
@@ -453,12 +453,13 @@ GLOBAL_LIST_INIT(potential_theft_objectives, (subtypesof(/datum/theft_objective)
 
 /datum/objective/download
 /datum/objective/download/proc/gen_amount_goal()
-	target_amount = rand(10,20)
-	explanation_text = "Download [target_amount] research levels."
+	target_amount = rand(20,40)
```
Comment (SabreML) [inline code/game/gamemodes/objective.dm:456]:
```suggestion
	target_amount = rand(20, 40)
```

**Comment by Kiyahitayika** [comment]
Someone asked a question about the hovering the icons. I can't seem to see that comment anymore. However, the answer is:

The code passes the "name" associated with each design item in a techweb to the TGUI interface. Hovering over each of the circuit board icons would indeed show a different description/tooltip.

```
		for(var/v in stored_research.researched_designs)
			var/datum/design/D = SSresearch.get_techweb_design_by_id(v)
			if(!(D.build_type & allowed_design_types) || !(selected_category in D.category) || length(D.reagents_list))
				continue
			var/list/design_out = list("id" = D.id, "name" = D.name, "cost" = get_design_cost(D), "time" = get_design_build_time(D))
```

File: code/controllers/subsystem/research.dm [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+		for(var/cr_path in subtypesof(/datum/chemical_reaction))
+			var/datum/chemical_reaction/CR = new cr_path()
+			if(CR.result in recursive_reagent_blacklist)
+				continue
+			for(var/reagent in CR.required_reagents)
+				// If one of our required reagents is in the first list
+				if(reagent in id_cache_1)
+					if(CR.result)
+						id_cache_2 |= CR.result
```
Comment (SabreML) [inline code/controllers/subsystem/research.dm:268]: Might as well, to save some space.
```suggestion
				if((reagent in id_cache_1) && CR.result)
					id_cache_2 |= CR.result
```

File: code/modules/research/chem_consumer.dm [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+		return
+
+	if(istype(I, /obj/item/reagent_containers/glass) || istype(I, /obj/item/reagent_containers/food/drinks))
+		if(panel_open)
+			to_chat(user, "<span class='notice'>Close the maintenance panel first.</span>")
+			return
+		if(!user.drop_item())
+			to_chat(user, "<span class='warning'>[I] is stuck to you!</span>")
+			return
+		beaker =  I
```
Comment (SabreML) [inline code/modules/research/chem_consumer.dm:57]:
```suggestion
		beaker = I
```

File: code/modules/research/chem_consumer.dm [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+		update_icon()
+
+/obj/machinery/research_chem_consumer/deconstruct()
+	if(beaker)
+		beaker.forceMove(get_turf(src))
+		beaker = null
+	return ..()
+
+/obj/machinery/research_chem_consumer/AltClick(mob/user)
+	if(!ishuman(usr))
```
Comment (SabreML) [inline code/modules/research/chem_consumer.dm:84]:
```suggestion
	if(!ishuman(user))
```

File: code/modules/research/chem_consumer.dm [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+		return
+	if(!istype(user) || user.incapacitated())
+		to_chat(user, "<span class='warning'>You can't do that right now!</span>")
+		return
+	if(!in_range(src, user))
```
Comment (SabreML) [inline code/modules/research/chem_consumer.dm:89]: `Adjacent()` instead, maybe?
```suggestion
	if(!Adjacent(user))
```

File: code/modules/research/chem_consumer.dm [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+		return
+
+	. = TRUE
+
+	switch(action)
+		if("eject_beaker")
+			eject_beaker()
+		if("consume")
+			var/reagent_amount = beaker.reagents.get_reagent_amount(SSresearch.complex_research_chem_id)
+			if(!(reagent_amount > 0))
```
Comment (SabreML) [inline code/modules/research/chem_consumer.dm:128]:
```suggestion
			if(!reagent_amount)
```

File: code/modules/research/rdconsole.dm [hunk 7]
```diff
@@ -681,9 +569,57 @@ won't update every console in existence) but it's more of a hassle to do. Also,
...
+			if(iscarbon(user))
+				var/obj/item/card/id/idcard = user.get_active_hand()
+				if(istype(idcard))
+					logname = "User: [idcard.registered_name]"
+			if(ishuman(user))
+				var/mob/living/carbon/human/H = user
+				var/obj/item/I = H.wear_id
+				if(istype(I))
+					var/obj/item/card/id/ID = I.GetID()
+					if(istype(ID))
```
Comment (SabreML) [inline code/modules/research/rdconsole.dm:613]: Always returns either the ID card or `null`.
```suggestion
					var/obj/item/card/id/ID = I.GetID()
					if(ID)
```

File: code/modules/research/server.dm [hunk 8]
```diff
@@ -2,359 +2,153 @@
...
-/obj/machinery/r_n_d/server/emp_act(severity)
-	griefProtection()
-	..()
-
-
-/obj/machinery/r_n_d/server/ex_act(severity)
-	griefProtection()
+/obj/machinery/r_n_d/server/emp_act()
+	stat |= EMPED
+	addtimer(CALLBACK(src, .proc/unemp), 600)
```
Comment (SabreML) [inline code/modules/research/server.dm:46]:
```suggestion
	addtimer(CALLBACK(src, .proc/unemp), 1 MINUTES)
```

File: code/modules/research/techwebs/__techweb_helpers.dm [hunk 9]
```diff
@@ -0,0 +1,88 @@
+/proc/count_unique_techweb_nodes()
+	var/static/list/L = typesof(/datum/techweb_node)
+	return L.len
+
+/proc/count_unique_techweb_designs()
+	var/static/list/L = typesof(/datum/design)
+	return L.len
```
Comment (SabreML) [inline code/modules/research/techwebs/__techweb_helpers.dm:7]:
```suggestion
/proc/count_unique_techweb_nodes()
	var/static/list/L = typesof(/datum/techweb_node)
	return length(L)

/proc/count_unique_techweb_designs()
	var/static/list/L = typesof(/datum/design)
	return length(L)
```

File: code/modules/research/techwebs/__techweb_helpers.dm [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+	var/list/all_techweb_products = list()
+
+	var/list/all_web_ids = list()
+
+	var/list/designs_without_webs = list()
+	var/list/invalid_web_designs = list()
+
+	for(var/x in subtypesof(/datum/design))
+		var/datum/design/D = new x
+		if(D.build_type in list(AUTOLATHE, BIOGENERATOR, SMELTER, (AUTOLATHE|SMELTER)))
```
Comment (SabreML) [inline code/modules/research/techwebs/__techweb_helpers.dm:52]:
```suggestion
		if(D.build_type in list(AUTOLATHE, BIOGENERATOR, SMELTER, (AUTOLATHE | SMELTER)))
```

File: code/modules/research/techwebs/_techweb.dm [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+	return unresearch_node(SSresearch.get_techweb_node_by_id(id))
+
+/datum/techweb/proc/unresearch_node(datum/techweb_node/node)
+	if(!istype(node))
+		return FALSE
+	researched_nodes -= node.id
+	recalculate_nodes(TRUE)				//Fully rebuild the tree.
+
+/datum/techweb/proc/boost_with_path(datum/techweb_node/N, itempath)
+	if(!istype(N)||!ispath(itempath))
```
Comment (SabreML) [inline code/modules/research/techwebs/_techweb.dm:150]:
```suggestion
	if(!istype(N) || !ispath(itempath))
```

File: code/modules/research/techwebs/_techweb.dm [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+			hidden_nodes -= N.id
+	return TRUE
+
+/datum/techweb/proc/update_node_status(datum/techweb_node/node)
+	var/researched = FALSE
+	var/available = FALSE
+	var/visible = FALSE
+	if(researched_nodes[node.id])
+		researched = TRUE
+	var/needed = node.prereq_ids.len
```
Comment (SabreML) [inline code/modules/research/techwebs/_techweb.dm:165]:
```suggestion
	var/needed = length(node.prereq_ids)
```

File: code/modules/research/techwebs/_techweb.dm [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+	if(researched)
+		researched_nodes[node.id] = node
+		for(var/i in node.designs)
+			add_design(node.designs[i])
+	else
+		if(available)
+			available_nodes[node.id] = node
+		else
+			if(visible)
+				visible_nodes[node.id] = node
```
Comment (SabreML) [inline code/modules/research/techwebs/_techweb.dm:186]:
```suggestion
		else if(visible)
			visible_nodes[node.id] = node
```

File: code/modules/research/chem_consumer.dm [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+	. = ..()
+	component_parts = list()
+	component_parts += new /obj/item/circuitboard/research_chem_consumer(null)
+
+/obj/machinery/research_chem_consumer/update_icon()
+	if(panel_open)
+		icon_state = "chemical_open"
+	else if(stat & NOPOWER)
+		icon_state = "chemical_powerless"
+	else if(istype(beaker))
```
Comment (SabreML) [inline code/modules/research/chem_consumer.dm:24]: Do you need the `istype()` here and on [L101](https://github.com/ParadiseSS13/Paradise/blob/fa553e08612a6f9729aa8da1078bd2c8894936f3/code/modules/research/chem_consumer.dm#L101)? It seems like if there's something inserted it'll always be a beaker.
```suggestion
	else if(beaker)
```

**Review by SabreML — COMMENTED** [review]
:eyes:

**Comment by Fox-McCloud** [comment]
How does TG handle the server ID's?

I do admit I have concerns of players being able to see the actual UID string of servers and such---I could...*hypothetically* see nonsense arising out of this.

**Comment by Sikroth** [comment]
Dept Lathes coming at a later date? What about resource silos? Good starting change regardless.

**Comment by Sinestia** [comment]
> 
> 
> > It turns science into something which actually requires thought
> 
> image

Tech webs are very flawed and mindless, in that they're basically just cookie clicker that people spam servers to do


It's still a marked upgrade over para's status quo of 5 minute research speedrun by throwing things into the deconstructor in the right order

**Comment by Butlerblock** [comment]
How would this affect balance for early game antags? Such as blob, wizard, or nukie, that crew definitely relies on rnd to be as fast as possible.(definitely not needed to win for wiz or blob, but heavily helps against nukies for things such as cargo points)

**Comment by Spacemanspark** [comment]
Miners can always get a stack of plasma in lavaland to gain cargo some points.

**Comment by SoThisIsItNow** [comment]
I played on servers with techwebs, its usually too slow to impact nukie rounds. Overall it makes rnd more mindless in my opinion. You click on the stuff you want after 5 minutes (Always mining tech, everything else would be silly) and then just wander off for 10 minutes and then just click on a random research again. The only thing people seem to do is cooling the server for more points and that is it, rnd is done.

**Comment by Necaladun** [comment]
@SoThisIsItNow Please be aware we don't give a single shit about your opinion, and the server staff actively want you to fuck off.

**Comment by Sinestia** [comment]
> 
> 
> I played on servers with techwebs, its usually too slow to impact nukie rounds. Overall it makes rnd more mindless in my opinion. You click on the stuff you want after 5 minutes (Always mining tech, everything else would be silly) and then just wander off for 10 minutes and then just click on a random research again. The only thing people seem to do is cooling the server for more points and that is it, rnd is done.

A poor meta isn't inherent to the concept. 

It can all be tuned (as has been done here seemingly.)

**Comment by Cocacolagua** [comment]
suggestion:

please let xenobiology or the experimentor contribute in some way to earn points, given that these are the only sub-departments left that don’t seem to have a way to help out

**Comment by Spacemanspark** [comment]
Xenobiology serves its own distinct purpose, separate from everything else. If it doesn't contribute that's fine imo.

The experimentor is... honestly... pointless. Personally, I wouldn't oppose it being removed.

**Comment by AffectedArc07** [comment]
> suggestion:
> 
> please let xenobiology or the experimentor contribute in some way to earn points, given that these are the only sub-departments left that don’t seem to have a way to help out
> 
> (other than that, good shit)

I have planned to add a slime core machine which can take in specific slime cores for research points. As for the experimentor, yeah that hasnt served a reasonable purpose since 2018 (You could use it to bump tech levels). I wouldnt cry if it got purged.

**Comment by MRgaty** [comment]
Oh god scichem and toxins being actually useful, I love this.

File: code/modules/research/chem_consumer.dm (hunk 5 above)
Comment (AffectedArc07) [reply code/modules/research/chem_consumer.dm:89]: `Adjacent` can recurse, so this is better

**Comment by AffectedArc07** [comment]
Whoever asked about hovering icons
image

**Comment by TDSSS** [comment]
Suggestion: Instead of circuitboards, show the machines those circuitboards are making.

**Comment by AffectedArc07** [comment]
> Suggestion: Instead of circuitboards, show the machines those circuitboards are making.

Not viable unless you think iterating through every single design, checking if it makes a circuit board, casting to that circuit board, looking up its target machine, generating an icon of said machine, then saving that. Not to mention computers wont generate properly anyway since their appearances are made of layers, not flat icons.

**Comment by TDSSS** [comment]
> > Suggestion: Instead of circuitboards, show the machines those circuitboards are making.
> 
> Not viable unless you think iterating through every single design, checking if it makes a circuit board, casting to that circuit board, looking up its target machine, generating an icon of said machine, then saving that. Not to mention computers wont generate properly anyway since their appearances are made of layers, not flat icons.

Couldn't we give designs a new var called something like icon_override. If present, techwebs displays that icon instead of the product's icon. If not present, it stays as it is now.

That could be made to work with computers as well.

**Comment by AffectedArc07** [comment]
> > > Suggestion: Instead of circuitboards, show the machines those circuitboards are making.
> > 
> > 
> > Not viable unless you think iterating through every single design, checking if it makes a circuit board, casting to that circuit board, looking up its target machine, generating an icon of said machine, then saving that. Not to mention computers wont generate properly anyway since their appearances are made of layers, not flat icons.
> 
> Couldn't we give designs a new var called something like icon_override. If present, techwebs displays that icon instead of the product's icon. If not present, it stays as it is now.
> 
> That could be made to work with computers as well.

Except it wouldnt work with consoles because they render as overlays, so unless you want to store every console screen as a duplicated sprite (Spoiler alert, thats not happening), it wont work.

**Comment by Sikroth** [comment]
Um, I dunno if it's worth the trouble especially if the implementation is super performance expensive (i am aware the techwebs UI can literally freeze the game for several seconds there, maybe this is related to that), but tg does show the constructed sprite for all the things you unlock. It might be worth looking into? I understand if it's too much trouble and it's not a big deal either way but it'd make those techweb nodes look a bit better and more intuitive at a glance.

image

**Comment by AffectedArc07** [comment]
@Sikroth I will see how TG handles it and attempt to replicate, but no promises.

**Review by Dumbdumn5 — APPROVED** [review]
Hopefully tech webs in this form do more than what people were worrying about having to sit around clicking boxes on occasion, especially with others being able to contribute to point gain in their own way. :+1:

**Comment by AffectedArc07** [comment]
I will return to this some day. Do not fear.

**Comment by AffectedArc07** [comment]
Closing this as I do not have the time or motivation to deal with it at the moment.

Anyone else is free to pick this up, regarding you follow repo code of conduct and get **maint** approval before opening the PR.

This is still a confirmed approval internally, since kyet's singular objection isn't enough to stunt a PR entirely. 

Again, only maint approval is required if you wish to re-open this, you do not need a head to be onboard.


All I will say is if you do pick this up, ensure you actually know how to port things properly, and not just copy paste TG code verbatim.

**Pull request closed without merging**